### PR TITLE
James 3502, add ssl support 

### PR DESCRIPTION
**Pull request by andrevka**
https://issues.apache.org/jira/browse/JAMES-3502?filter=-2

### CODE REVIEW

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java [hunk 1]
```diff
@@ -18,18 +18,278 @@
...
 import java.util.Optional;
+import java.util.stream.Stream;
 
 import org.apache.commons.configuration2.Configuration;
 
 import com.google.common.base.Preconditions;
 import com.google.common.base.Strings;
 
+import static org.apache.james.backends.rabbitmq.RabbitMQConfiguration.SSLConfiguration.*;
+import static org.apache.james.backends.rabbitmq.RabbitMQConfiguration.SSLConfiguration.SSLValidationStrategy.*;
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java:36]: Please avoid wildcards, this will likely not pass the checkstyle.

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java [hunk 2]
```diff
@@ -18,18 +18,278 @@
...
+
+            @Override
+            public final int hashCode() {
+                return Objects.hash(file, Arrays.hashCode(password));
+            }
+        }
+
+        static class Builder {
+
+            interface RequireSSLStrategyTrustStore {
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java:170]: Use `@FunctionalInterface` annotation where relevant.

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java [hunk 3]
```diff
@@ -18,18 +18,278 @@
...
+                return Stream.of(values())
+                        .filter(strategy -> strategy.name().equalsIgnoreCase(rawValue))
+                        .findAny()
+                        .orElseThrow(() -> new IllegalArgumentException(String.format("invalid strategy '%s'", rawValue)));
+
+            }
+        }
+
+        public enum HostNameVerifier {
+            DEFAULT,
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java:59]: DEFAUT does not describe a behaviour. Can we find terms to express what this does?

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConnectionFactory.java [hunk 4]
```diff
@@ -54,18 +69,85 @@ private ConnectionFactory from(RabbitMQConfiguration rabbitMQConfiguration) {
             connectionFactory.setUsername(rabbitMQConfiguration.getManagementCredentials().getUser());
             connectionFactory.setPassword(String.valueOf(rabbitMQConfiguration.getManagementCredentials().getPassword()));
 
+            if (configuration.useSsl()) setupSslConfiguration(connectionFactory);
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConnectionFactory.java:72]: You miss a `{}` block after this if, the code style will not like it ;-)

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc [hunk 5]
```diff
@@ -46,6 +46,39 @@ Optional integer, defaults to 50
...
+| Configure host name verification. Possible options are default and accept_any_hostname
+Optional string, defaults to default
+
+| ssl.keystore
+| Points to the keystore(PKCS12) used for verifying rabbitmq connection. If configured then "ssl.keystore.password" must also be configured,
+Optional string, defaults to empty string
+
+| ssl.keystore.password
+| Configure the keystore password. If configured then "ssl.keystore" must also be configured,
+Optional string, defaults to empty string
```
Comment (chibenwa) [inline docs/modules/servers/pages/distributed/configure/rabbitmq.adoc:81]: Is there some cases when it should be specified?

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc [hunk 6]
```diff
@@ -46,6 +46,39 @@ Optional integer, defaults to 50
...
+| Configure the truststore password. If configured then "ssl.truststore" must also be configured,
+Optional string, defaults to empty string. "ssl.validation.strategy: override" must be configured if you want to use this
+
+| ssl.hostname.verifier
+| Configure host name verification. Possible options are default and accept_any_hostname
+Optional string, defaults to default
+
+| ssl.keystore
+| Points to the keystore(PKCS12) used for verifying rabbitmq connection. If configured then "ssl.keystore.password" must also be configured,
+Optional string, defaults to empty string
```
Comment (chibenwa) [inline docs/modules/servers/pages/distributed/configure/rabbitmq.adoc:77]: Is there some cases when it should be specified?

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc [hunk 7]
```diff
@@ -46,6 +46,39 @@ Optional integer, defaults to 50
...
+| Points to the truststore (PKCS12) used for verifying rabbitmq connection. If configured then "ssl.truststore.password" must also be configured,
+Optional string, defaults to systemwide truststore. "ssl.validation.strategy: override" must be configured if you want to use this
+
+| ssl.truststore.password
+| Configure the truststore password. If configured then "ssl.truststore" must also be configured,
+Optional string, defaults to empty string. "ssl.validation.strategy: override" must be configured if you want to use this
+
+| ssl.hostname.verifier
+| Configure host name verification. Possible options are default and accept_any_hostname
+Optional string, defaults to default
```
Comment (chibenwa) [inline docs/modules/servers/pages/distributed/configure/rabbitmq.adoc:73]: What does default do?

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc [hunk 8]
```diff
@@ -46,6 +46,39 @@ Optional integer, defaults to 50
...
+| Is using ssl enabled
+Optional boolean, defaults to false
+
+| ssl.management.enabled
+| Is using ssl on management api enabled
+Optional boolean, defaults to false
+
+| ssl.validation.strategy
+| Configure the validation strategy used for rabbitmq connections. Possible values are default, ignore and override.
+Optional string, defaults to default
```
Comment (chibenwa) [inline docs/modules/servers/pages/distributed/configure/rabbitmq.adoc:61]: What does default do?

**Review by chibenwa — COMMENTED** [review]
Nice work!

Please don't forget to update `src/site/xdoc/server/config-rabbitmq.xml` as well!

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc (hunk 5 above)
Comment (andrevka) [reply docs/modules/servers/pages/distributed/configure/rabbitmq.adoc:81]: client certificate authentication?

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc (hunk 6 above)
Comment (andrevka) [reply docs/modules/servers/pages/distributed/configure/rabbitmq.adoc:77]: client certificate authentication?

File: backends-common/rabbitmq/pom.xml [hunk 9]
```diff
@@ -93,6 +93,16 @@
...
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+            <version>4.5.13</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpcore</artifactId>
+            <version>4.4.5</version>
+        </dependency>
```
Comment (chibenwa) [inline backends-common/rabbitmq/pom.xml:105]: Can you quickly explain why we need these dependencies? At first glance it seems unrelated to AMQP...

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc (hunk 6 above)
Comment (chibenwa) [reply docs/modules/servers/pages/distributed/configure/rabbitmq.adoc:77]: Sorry, I missread.

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc (hunk 5 above)
Comment (chibenwa) [reply docs/modules/servers/pages/distributed/configure/rabbitmq.adoc:81]: Sorry, I missread. That's ok.

File: backends-common/rabbitmq/pom.xml (hunk 9 above)
Comment (andrevka) [reply backends-common/rabbitmq/pom.xml:105]: httpcore, has useful class SSLContextBuilder
httpclient, is needed for configuring mangement api client for org.apache.http.conn.ssl.DefaultHostnameVerifier

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java (hunk 3 above)
Comment (andrevka) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java:63]: SUBJECT_ALTERNATIVE_NAME_VERIFIER ?

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc (hunk 7 above)
Comment (andrevka) [reply docs/modules/servers/pages/distributed/configure/rabbitmq.adoc]: fixed https://github.com/apache/james-project/pull/307/commits/3c0c552a7fe1918e07d9fe305ede2d1552e95813

File: docs/modules/servers/pages/distributed/configure/rabbitmq.adoc (hunk 8 above)
Comment (andrevka) [reply docs/modules/servers/pages/distributed/configure/rabbitmq.adoc]: fixed https://github.com/apache/james-project/pull/307/commits/3c0c552a7fe1918e07d9fe305ede2d1552e95813

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java (hunk 2 above)
Comment (andrevka) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java:175]: https://github.com/apache/james-project/pull/307/commits/62bd3eedb8553408878ee74e59425ae7e1dabee5

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java (hunk 1 above)
Comment (andrevka) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java]: removed, applied checkstyle

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Would other contributors like to have a look at this work?

It is :green_apple: .

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java [hunk 10]
```diff
@@ -197,11 +471,16 @@ public static RabbitMQConfiguration from(Configuration configuration) {
...
+        Boolean useSsl = configuration.getBoolean(USE_SSL, false);
+
         ManagementCredentials managementCredentials = ManagementCredentials.from(configuration);
         return builder()
             .amqpUri(amqpUri)
             .managementUri(managementUri)
             .managementCredentials(managementCredentials)
+            .useSsl(useSsl)
+            .sslConfiguration(sslTrustConfiguration(configuration))
+            .sslConfiguration(sslTrustConfiguration(configuration))
```
Comment (Arsnael) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQConfiguration.java:502]: duplicated line here

File: backends-common/rabbitmq/pom.xml [hunk 11]
```diff
@@ -93,6 +93,16 @@
             <groupId>org.apache.commons</groupId>
             <artifactId>commons-pool2</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+            <version>4.5.13</version>
```
Comment (Arsnael) [inline backends-common/rabbitmq/pom.xml:99]: We already have that artifact with the same version declared in the root pom file, so you can omit the version here

File: backends-common/rabbitmq/pom.xml [hunk 12]
```diff
@@ -93,6 +93,16 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpcore</artifactId>
+            <version>4.4.5</version>
```
Comment (Arsnael) [inline backends-common/rabbitmq/pom.xml:104]: I can see we have an httpcore artifact declared in `james-server-jmap` module as well, in a different version (4.4.14). Would be good to try centralizing the use of the artifact in the root pom and bump it for jmap

**Review by Arsnael — COMMENTED** [review]
Good job overall thanks, just a few minor comments on my part :)

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged
